File: 129-sum-root-to-leaf-numbers/sum-root-to-leaf-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumDownwards(TreeNode* root, int total){
        total = total*10 + root->val;
        if (root->left == NULL && root->right == NULL){
            return total;
        }

        if (root->left == NULL){
            return sumDownwards(root->right, total);
        }else if (root->right == NULL){
            return sumDownwards(root->left, total);
        }

        return sumDownwards(root->left, total) + sumDownwards(root->right, total);
    }

    int sumNumbers(TreeNode* root) {
        return sumDownwards(root, 0);
    }

};

File: 129-sum-root-to-leaf-numbers/sum-root-to-leaf-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumPaths(self, root, total):
        total = total*10 + root.val

        if ((root.left == None) and (root.right == None)):
            return total
        
        children_totals = 0
        if (root.left != None):
            children_totals += self.sumPaths(root.left, total)

        if (root.right != None):
            children_totals += self.sumPaths(root.right, total)

        return children_totals

    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        return self.sumPaths(root, 0)
        